pkg/log/log_test.go

```diff
@@ -0,0 +1,528 @@
+	test := NewLogger("test")
+
+	debugSignal := syscall.SIGUSR1
+	SetupDebugToggleSignal(debugSignal)
+	sigCh := make(chan os.Signal, 1)
+	signal.Notify(sigCh, debugSignal)
+	flag.Set(optDebug, "off:*")
+	debugging := false
+
+	expected := []string{}
+	messages := []string{"debug", "info", "warning", "error"}
+	for i := 0; i < 2; i++ {
+		for _, msg := range messages {
+			var logfn func(string, ...interface{})
+
+			filtered := false
+			switch msg {
+			case "debug":
+				logfn = test.Debug
+				filtered = !debugging
+			case "info":
+				logfn = test.Info
+			case "warning":
+				logfn = test.Warn
+			case "error":
+				logfn = test.Error
+			default:
+				continue
+			}
+			logfn("%s", msg)
+			if !filtered {
+				expected = append(expected, msg)
+			}
+		}
+		log.forceDebug(!log.debugForced())
+		debugging = !debugging
+	}
+
+	sources := map[string]struct{}{
+		"test": {},
+	}
+
+	tl.check(expected, true, sources)
+}
+
+func getenv(key string, fallback interface{}) interface{} {
+	strval := os.Getenv(key)
+	if strval == "" {
+		return fallback
+	}
+	switch defv := fallback.(type) {
+	case int:
+		v, err := strconv.ParseInt(strval, 10, 0)
+		if err != nil {
+			fmt.Printf("error: invalid environment variable %s = %s: %v\n", key, strval, err)
+			return defv
+		}
+		return int(v)
+	case time.Duration:
+		v, err := time.ParseDuration(strval)
+		if err != nil {
+			fmt.Printf("error: invalid enviroment variable %s = %s: %v\n", key, strval, err)
+			return defv
+		}
+		return v
+	case []bool:
+		v := []bool{}
+		for _, strb := range strings.Split(strval, ",") {
+			b, err := strconv.ParseBool(strb)
+			if err != nil {
+				fmt.Printf("error: invalid environment variable %s = %s: %v\n", key, strval, err)
+				return defv
+			}
+			v = append(v, b)
+		}
+		return v
+
+	default:
+		panic(fmt.Sprintf("enviroment variable %s=%s with unhandled type %T",
+			key, strval, fallback))
+	}
+}
+
+var (
+	// number of concurrent loggers, togglers, test duration, test verbosity
+	numLoggers    = getenv("LOGTEST_LOGGERS", 32).(int)
+	numTogglers   = getenv("LOGTEST_TOGGLERS", 4).(int)
+	testDuration  = getenv("LOGTEST_DURATION", 15*time.Second).(time.Duration)
+	testVerbosity = getenv("LOGTEST_VERBOSITY", []bool{true, false}).([]bool)
+)
+
+// createLoggers creates a bunch of loggers.
+func createLoggers(cnt int) []Logger {
+	loggers := make([]Logger, cnt, cnt)
+	for idx := range loggers {
+		switch idx % 7 {
+		case 0:
+			loggers[idx] = NewLogger(fmt.Sprintf("%d-<0>-test", idx))
+		case 1:
+			loggers[idx] = NewLogger(fmt.Sprintf("%d-<1>-logger", idx))
+		case 2:
+			loggers[idx] = NewLogger(fmt.Sprintf("%d-<2>-logging", idx))
+		case 3:
+			loggers[idx] = NewLogger(fmt.Sprintf("%d-<3>-log", idx))
+		case 4:
+			loggers[idx] = NewLogger(fmt.Sprintf("%d-<4>-logger", idx))
+		case 5:
+			loggers[idx] = NewLogger(fmt.Sprintf("%d-<5>-tester", idx))
+		default:
+			loggers[idx] = NewLogger(fmt.Sprintf("%d-<->-logger-tester", idx))
+		}
+
+		if (idx % 11) == 0 {
```

Comment:
Maybe a comment for explaining the indexing would be nice. :-)

Reply:
Added a clarifying comment.